### PR TITLE
Laboratory work #3, Isakov Danila - 21FPL2

### PR DESCRIPTION
**Pull request by vchemsmisl**

### CODE REVIEW

File: lab_3_keywords_textrank/main.py [hunk 1]
```diff
@@ -443,14 +533,23 @@ def fill_positions(self, tokens: tuple[int, ...]) -> None:
             tokens : tuple[int, ...]
                 sequence of tokens
         """
-        pass
+        for token in tokens:
+            if token not in self._positions:
+                token_positions = [idx + 1 for idx, elem in enumerate(tokens) if elem == token]
```
Comment (artyomtugaryov) [inline lab_3_keywords_textrank/main.py:538]: enumerate(tokens, start=1):

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]

File: lab_3_keywords_textrank/main.py [hunk 2]
```diff
@@ -36,7 +38,8 @@ def __init__(self, stop_words: tuple[str, ...], punctuation: tuple[str, ...]) ->
             punctuation : tuple[str, ...]
                 punctuation symbols to remove during text cleaning
         """
-        pass
+        self.stop_words = stop_words
```
[the remark this answers is not in the document]
Comment (vchemsmisl) [reply lab_3_keywords_textrank/main.py:41]: Done

File: lab_3_keywords_textrank/main.py [hunk 3]
```diff
@@ -51,7 +54,7 @@ def _clean_and_tokenize(self, text: str) -> tuple[str, ...]:
             tuple[str, ...]
                 clean lowercase tokens
         """
-        pass
+        return tuple(''.join([symbol.lower() for symbol in text if symbol not in self.punctuation]).split())
```
[the remark this answers is not in the document]
Comment (vchemsmisl) [reply lab_3_keywords_textrank/main.py:57]: Done

File: lab_3_keywords_textrank/main.py [hunk 4]
```diff
@@ -120,7 +125,9 @@ def _learn_indices(self, tokens: tuple[str, ...]) -> None:
             tokens : tuple[str, ...]
                 sequence of string tokens
         """
-        pass
+        self._word2id = {token: randint(1000, 9999) for token in tokens}
+        for index, token in enumerate(self._word2id.keys()):
```
[the remark this answers is not in the document]
Comment (vchemsmisl) [reply lab_3_keywords_textrank/main.py:129]: Done

File: lab_3_keywords_textrank/main.py [hunk 5]
```diff
@@ -403,7 +484,12 @@ def is_incidental(self, vertex1: int, vertex2: int) -> int:
                 1 if vertices are incidental, otherwise 0
         If either of vertices is not present in the graph, -1 is returned
         """
-        pass
+        if vertex1 not in self._edges or vertex2 not in self._edges:
+            return -1
+        if vertex2 in self._edges[vertex1]:
+            return 1
+        else:
```
[the remark this answers is not in the document]
Comment (vchemsmisl) [reply lab_3_keywords_textrank/main.py:491]: Done

File: lab_3_keywords_textrank/main.py (hunk 1 above)
Comment (vchemsmisl) [reply lab_3_keywords_textrank/main.py:538]: Done

File: lab_3_keywords_textrank/main.py [hunk 6]
```diff
@@ -120,7 +130,11 @@ def _learn_indices(self, tokens: tuple[str, ...]) -> None:
             tokens : tuple[str, ...]
                 sequence of string tokens
         """
-        pass
+        self._word2id = {token: randint(1000, 9999) for token in tokens}
```
Comment (artyomtugaryov) [inline lab_3_keywords_textrank/main.py:133]: for different tokens you can get equals ints

File: lab_3_keywords_textrank/main.py [hunk 7]
```diff
@@ -120,7 +130,11 @@ def _learn_indices(self, tokens: tuple[str, ...]) -> None:
...
+        for index, token in enumerate(self._word2id):
+            w2i_values = list(self._word2id.values())
+            needed_value = w2i_values[index]
```
Comment (artyomtugaryov) [inline lab_3_keywords_textrank/main.py:136]: why do not you use `self._word2id[token]` ?

File: lab_3_keywords_textrank/main.py [hunk 8]
```diff
@@ -233,7 +273,25 @@ def add_edge(self, vertex1: int, vertex2: int) -> int:
                 0 if edge was added successfully, otherwise -1
         In case of vertex1 being equal to vertex2, -1 is returned as loops are prohibited
         """
-        pass
+        if vertex1 == vertex2:
+            return -1
+        for vertex in [vertex1, vertex2]:
+            if not vertex in self._matrix[0]:
+                self._matrix[0].append(vertex)
```
Comment (artyomtugaryov) [inline lab_3_keywords_textrank/main.py:280]: do not use self._matrix[0] to store vertexes

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]

File: lab_3_keywords_textrank/main.py (hunk 6 above)
Comment (vchemsmisl) [reply lab_3_keywords_textrank/main.py:133]: Done

File: lab_3_keywords_textrank/main.py (hunk 7 above)
Comment (vchemsmisl) [reply lab_3_keywords_textrank/main.py:136]: Done

File: lab_3_keywords_textrank/main.py (hunk 8 above)
Comment (vchemsmisl) [reply lab_3_keywords_textrank/main.py:280]: Done

File: lab_3_keywords_textrank/main.py [hunk 9]
```diff
@@ -120,7 +129,9 @@ def _learn_indices(self, tokens: tuple[str, ...]) -> None:
             tokens : tuple[str, ...]
                 sequence of string tokens
         """
-        pass
+        self._word2id = {token: 1000 + index for index, token in enumerate(tokens)}
```
Comment (artyomtugaryov) [inline lab_3_keywords_textrank/main.py:132]: looks like you can replace value for a token with a new value. It is not a problem, but better enumerate by set of tokens

Comment (vchemsmisl) [reply lab_3_keywords_textrank/main.py]: Done

**Review by artyomtugaryov — APPROVED** [review]

**Pull request closed without merging**